text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1087 @@
+`const fn foo(f: (const) fn()) { f() }`.
+
+These require changing the type system, making the constness of a function pointer part of the type.
+This in turn implies that a `const fn()` function pointer, a `(const) fn()` function pointer and a `fn()` function pointer could have
+different `TypeId`s, which is something that requires more design and consideration to clarify whether supporting downcasting with `Any`
+or just supporting `TypeId` equality checks detecting constness is desirable.
+
+Furthermore `const fn()` pointers introduce a new situation: you can transmute arbitrary values (e.g. null pointers, or just integers) to
+`const fn()` pointers, and the type system will not protect you. Instead the const evaluator will reject that when it actually
+evaluateds the code around the function pointer or even as late as when the function call happens.
+
+## `const` closures
+
+Closures need explicit opt-in to be callable in const contexts.
+You can already use closures in const contexts today to e.g. declare consts of function pointer type.
+So what we additionally need is some syntax like `const || {}` to declare a closure that implements
+`const Fn()`. See also [this tracking issue](https://github.com/rust-lang/project-const-traits/issues/10)
+While it may seem tempting to just automatically implement `const Fn()` (or `(const) Fn()`) where applicable,
+it's not clear that this can be done, and there are definite situations where it can't be done.
+As further experimentation is needed here, const closures are not part of this RFC.
+
+## Allow impls to refine any trait's methods
+
+We could allow writing `const fn` in impls without the trait opting into it.
+This would not affect `T: Trait` bounds, but still allow non-generic calls.
+
+This is similar to other refinings in impls, as the function still satisfies everything from the trait.
+
+Example: without adjusting `rand` for const trait support at all, users could write
+
+```rust
+struct CountingRng(u64);
+
+impl RngCore for CountingRng {
+    const fn next_u32(&mut self) -> u32 {
+        self.next_u64() as u32
+    }
+
+    const fn next_u64(&mut self) -> u64 {
+        self.0 += 1;
+        self.0
+    }
+
+    const fn fill_bytes(&mut self, dest: &mut [u8]) {
+        let mut left = dest;
+        while left.len() >= 8 {
+            let (l, r) = { left }.split_at_mut(8);
+            left = r;
+            let chunk: [u8; 8] = rng.next_u64().to_le_bytes();
+            l.copy_from_slice(&chunk);
+        }
+        let n = left.len();
+        let chunk: [u8; 8] = rng.next_u64().to_le_bytes();
+        left.copy_from_slice(&chunk[..n]);
+    }
+
+    const fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        Ok(self.fill_bytes(dest))
+    }
+}
+```
+
+and use it in non-generic code.
+It is not clear this is doable soundly for generic methods.
```

Comment:
It’s just a future possibility, so we can always figure out these details later, but I still don’t understand why this might not be possible to do. Is it just an implementation difficulty concern?